js-library/src/services/ens.ts

```diff
@@ -94,46 +102,84 @@ export class ENS {
   }
 
   /**
-   * Sets owner of the provided username on ENS
+   * Creates a request object that needs to be used to invoke the registerUsername method.
+   * The request object can be used to invoke the method multiple times without making any
+   * inconsistencies. This can be useful if registration fails because of insufficient funds,
+   * or some other reason.
    * @param username ENS username
    * @param address Owner address of the username
    * @param publicKey Hex string of a public key
+   * @returns ServiceRequest instance that can be used to invoke the request multiple times if fails.
    */
-  public async registerUsername(
+  public createRegisterUsernameRequest(
     username: Username,
     address: EthAddress,
     publicKey: PublicKey,
     expires: number = 86400,
-  ): Promise<void> {
+  ): ServiceRequest<RegisterUsernameRequestData> {
+    const request: ServiceRequest<RegisterUsernameRequestData> = {
+      stage: 0,
+      data: {
+        username,
+        address,
+        publicKey,
+        expires,
+      },
+      completedTxs: [],
+    }
+
+    return request
+  }
+
+  /**
+   * Sets owner of the provided username on ENS
+   * @param registerRequest request object previously created by the createRegisterUsernameRequest method
+   */
+  public async registerUsername(registerRequest: ServiceRequest<RegisterUsernameRequestData>): Promise<void> {
+    const {
+      data: { username, address, publicKey, expires },
+    } = registerRequest
     try {
-      assertUsername(username)
+      if (registerRequest.stage === 0) {
```

Comment:
it was good to give stage names instead of numbers, since studying the code from the outside it is not entirely clear what exactly this stage means.

Reply:
Maybe we could find some more easier way to use the stage parameter, but converting it to string might make code more complex. Because the checks will have to use OR/AND. This way it's just a simple greater than operator.

I added description to the Request interface which should make the code more readable.

Reply:
Actually, maybe the best way is to use enum with integers for stages.